File: android/app/src/main/java/com/locationsharing/app/ui/friends/FriendsListViewModel.kt
package com.locationsharing.app.ui.friends

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.locationsharing.app.data.friends.Friend
import com.locationsharing.app.data.friends.FriendsRepository
import com.locationsharing.app.data.friends.RealTimeFriendsService
import com.locationsharing.app.data.friends.ConnectionState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.catch
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.launch
import timber.log.Timber
import javax.inject.Inject

/**
 * ViewModel for the Friends List screen with real-time Firebase integration
 * Handles friends data, status updates, and UI state management
 */
@HiltViewModel
class FriendsListViewModel @Inject constructor(
    private val friendsRepository: FriendsRepository,
    private val realTimeFriendsService: RealTimeFriendsService
) : ViewModel() {
    
    private val _uiState = MutableStateFlow(FriendsListUiState())
    val uiState: StateFlow<FriendsListUiState> = _uiState.asStateFlow()
    
    private val _connectionState = MutableStateFlow(ConnectionState.DISCONNECTED)
    val connectionState: StateFlow<ConnectionState> = _connectionState.asStateFlow()
    
    init {
        startRealTimeSync()
        observeFriendsData()
        observeConnectionState()
    }
    
    /**
     * Start real-time synchronization with Firebase
     */
    private fun startRealTimeSync() {
        viewModelScope.launch {
            try {
                realTimeFriendsService.startSync()
                Timber.d("Real-time friends sync started")
            } catch (e: Exception) {
                Timber.e(e, "Error starting real-time sync")
                _uiState.value = _uiState.value.copy(
                    error = "Failed to connect to real-time updates: ${e.message}",
                    isLoading = false
                )
            }
        }
    }
    
    /**
     * Observe friends data from Firebase with real-time updates
     */
    private fun observeFriendsData() {
        viewModelScope.launch {
            combine(
                friendsRepository.getFriends(),
                friendsRepository.getOnlineFriends()
            ) { allFriends, onlineFriends ->
                Pair(allFriends, onlineFriends)
            }
            .catch { exception ->
                Timber.e(exception, "Error observing friends data")
                _uiState.value = _uiState.value.copy(
                    error = "Failed to load friends: ${exception.message}",
                    isLoading = false
                )
            }
            .collect { (allFriends, onlineFriends) ->
                _uiState.value = _uiState.value.copy(
                    friends = allFriends,
                    onlineFriends = onlineFriends,
                    isLoading = false,
                    error = null,
                    lastUpdateTime = System.currentTimeMillis()
                )
                
                Timber.d("Friends updated: ${allFriends.size} total, ${onlineFriends.size} online")
            }
        }
    }
    
    /**
     * Observe connection state changes
     */
    private fun observeConnectionState() {
        viewModelScope.launch {
            realTimeFriendsService.connectionState.collect { state ->
                _connectionState.value = state
                
                _uiState.value = _uiState.value.copy(
                    isConnected = state == ConnectionState.CONNECTED,
                    isLoading = state == ConnectionState.CONNECTING
                )
                
                // Show connection status messages
                when (state) {
                    ConnectionState.ERROR -> {
                        _uiState.value = _uiState.value.copy(
                            error = "Connection lost. Retrying..."
                        )
                    }
                    ConnectionState.CONNECTED -> {
                        if (_uiState.value.lastUpdateTime > 0) {
                            // Only show success message if we had data before
                            _uiState.value = _uiState.value.copy(
                                successMessage = "Connected to real-time updates"
                            )
                        }
                    }
                    else -> { /* Handle other states if needed */ }
                }
                
                Timber.d("Connection state changed: $state")
            }
        }
    }
    
    /**
     * Refresh friends data manually
     */
    fun refreshFriends() {
        viewModelScope.launch {
            try {
                _uiState.value = _uiState.value.copy(
                    isRefreshing = true,
                    error = null
                )
                
                // Restart real-time sync to refresh data
                realTimeFriendsService.stopSync()
                kotlinx.coroutines.delay(500) // Brief pause
                realTimeFriendsService.startSync()
                
                _uiState.value = _uiState.value.copy(
                    isRefreshing = false,
                    successMessage = "Friends list refreshed"
                )
                
                Timber.d("Friends data refreshed manually")
            } catch (e: Exception) {
                Timber.e(e, "Error refreshing friends")
                _uiState.value = _uiState.value.copy(
                    isRefreshing = false,
                    error = "Failed to refresh friends: ${e.message}"
                )
            }
        }
    }
    
    /**
     * Handle friend selection
     */
    fun selectFriend(friend: Friend) {
        _uiState.value = _uiState.value.copy(selectedFriend = friend)
        Timber.d("Friend selected: ${friend.name}")
    }
    
    /**
     * Clear friend selection
     */
    fun clearFriendSelection() {
        _uiState.value = _uiState.value.copy(selectedFriend = null)
        Timber.d("Friend selection cleared")
    }
    
    /**
     * Send friend request
     */
    suspend fun sendFriendRequest(toUserId: String, message: String? = null): Result<String> {
        return try {
            friendsRepository.sendFriendRequest(toUserId, message)
        } catch (e: Exception) {
            Timber.e(e, "Error sending friend request")
            Result.failure(e)
        }
    }
    
    /**
     * Remove friend with confirmation
     */
    suspend fun removeFriend(friendId: String): Result<Unit> {
        return try {
            val result = friendsRepository.removeFriend(friendId)
            
            if (result.isSuccess) {
                // Clear selection if removed friend was selected
                if (_uiState.value.selectedFriend?.id == friendId) {
                    clearFriendSelection()
                }
                
                _uiState.value = _uiState.value.copy(
                    successMessage = "Friend removed successfully"
                )
            }
            
            result
        } catch (e: Exception) {
            Timber.e(e, "Error removing friend")
            Result.failure(e)
        }
    }
    
    /**
     * Get friend by ID
     */
    fun getFriendById(id: String): Friend? {
        return _uiState.value.friends.find { it.id == id }
    }
    
    /**
     * Check if friend is currently online
     */
    fun isFriendOnline(friendId: String): Boolean {
        return getFriendById(friendId)?.isOnline() ?: false
    }
    
    /**
     * Clear error message
     */
    fun clearError() {
        _uiState.value = _uiState.value.copy(error = null)
    }
    
    /**
     * Clear success message
     */
    fun clearSuccessMessage() {
        _uiState.value = _uiState.value.copy(successMessage = null)
    }
    
    /**
     * Handle network connectivity changes
     */
    fun onNetworkConnectivityChanged(isConnected: Boolean) {
        if (isConnected) {
            // Restart sync when network is restored
            startRealTimeSync()
        } else {
            _uiState.value = _uiState.value.copy(
                error = "No internet connection. Some features may not work."
            )
        }
    }
    
    /**
     * Handle app lifecycle changes
     */
    fun onAppResumed() {
        // Refresh data when app comes to foreground
        refreshFriends()
    }
    
    fun onAppPaused() {
        // Optionally pause real-time updates to save battery
        // realTimeFriendsService.pauseSync()
    }
    
    override fun onCleared() {
        super.onCleared()
        // Stop real-time sync when ViewModel is cleared
        realTimeFriendsService.stopSync()
        Timber.d("FriendsListViewModel cleared, real-time sync stopped")
    }
}

/**
 * UI state for the Friends List screen
 */
data class FriendsListUiState(
    val friends: List<Friend> = emptyList(),
    val onlineFriends: List<Friend> = emptyList(),
    val selectedFriend: Friend? = null,
    val isLoading: Boolean = true,
    val isRefreshing: Boolean = false,
    val isConnected: Boolean = false,
    val error: String? = null,
    val successMessage: String? = null,
    val lastUpdateTime: Long = 0L
) {
    val hasOnlineFriends: Boolean
        get() = onlineFriends.isNotEmpty()
    
    val hasAnyFriends: Boolean
        get() = friends.isNotEmpty()
    
    val offlineFriends: List<Friend>
        get() = friends.filter { !it.isOnline() }
    
    val friendsCount: Int
        get() = friends.size
    
    val onlineFriendsCount: Int
        get() = onlineFriends.size
    
    val offlineFriendsCount: Int
        get() = offlineFriends.size
}

File: ios/LocationSharingApp/LocationSharingApp/Presentation/Friends/FriendsListViewModel.swift
import Foundation
import Combine
import SwiftUI

/**
 * ViewModel for the Friends List view with real-time Firebase integration
 * Handles friends data, status updates, and UI state management
 */
@MainActor
class FriendsListViewModel: ObservableObject {
    
    // MARK: - Published Properties
    
    @Published var friends: [Friend] = []
    @Published var onlineFriends: [Friend] = []
    @Published var selectedFriend: Friend?
    @Published var isLoading = true
    @Published var isRefreshing = false
    @Published var isConnected = false
    @Published var error: String?
    @Published var successMessage: String?
    
    // MARK: - Private Properties
    
    private let friendsRepository: FriendsRepository
    private var cancellables = Set<AnyCancellable>()
    private var lastUpdateTime: Date = Date()
    
    // MARK: - Computed Properties
    
    var hasOnlineFriends: Bool {
        !onlineFriends.isEmpty
    }
    
    var hasAnyFriends: Bool {
        !friends.isEmpty
    }
    
    var offlineFriends: [Friend] {
        friends.filter { !$0.isOnline }
    }
    
    var friendsCount: Int {
        friends.count
    }
    
    var onlineFriendsCount: Int {
        onlineFriends.count
    }
    
    var offlineFriendsCount: Int {
        offlineFriends.count
    }
    
    // MARK: - Initialization
    
    init(friendsRepository: FriendsRepository = DIContainer.shared.friendsRepository) {
        self.friendsRepository = friendsRepository
        setupObservers()
    }
    
    // MARK: - Public Methods
    
    /**
     * Start real-time synchronization with Firebase
     */
    func startRealTimeSync() {
        isLoading = true
        error = nil
        observeFriendsData()
    }
    
    /**
     * Stop real-time synchronization
     */
    func stopRealTimeSync() {
        cancellables.removeAll()
    }
    
    /**
     * Refresh friends data manually
     */
    func refreshFriends() {
        isRefreshing = true
        error = nil
        
        // Restart observers to refresh data
        cancellables.removeAll()
        observeFriendsData()
        
        // Simulate refresh completion after a short delay
        DispatchQueue.main.asyncAfter(deadline: .now() + 1.0) {
            self.isRefreshing = false
            self.successMessage = "Friends list refreshed"
            
            // Clear success message after 3 seconds
            DispatchQueue.main.asyncAfter(deadline: .now() + 3.0) {
                self.successMessage = nil
            }
        }
    }
    
    /**
     * Handle friend selection
     */
    func selectFriend(_ friend: Friend) {
        selectedFriend = friend
        print("Friend selected: \(friend.user.name)")
    }
    
    /**
     * Clear friend selection
     */
    func clearFriendSelection() {
        selectedFriend = nil
        print("Friend selection cleared")
    }
    
    /**
     * Send friend request
     */
    func sendFriendRequest(toUserId: String, message: String? = nil) async -> Result<String, Error> {
        // Implementation would depend on your FriendsRepository interface
        // For now, return a placeholder
        return .failure(NSError(domain: "NotImplemented", code: 0, userInfo: [NSLocalizedDescriptionKey: "Not implemented"]))
    }
    
    /**
     * Remove friend with confirmation
     */
    func removeFriend(_ friendId: String) async -> Result<Void, Error> {
        do {
            // Implementation would depend on your FriendsRepository interface
            // For now, simulate success
            
            // Clear selection if removed friend was selected
            if selectedFriend?.id == friendId {
                clearFriendSelection()
            }
            
            successMessage = "Friend removed successfully"
            
            // Clear success message after 3 seconds
            DispatchQueue.main.asyncAfter(deadline: .now() + 3.0) {
                self.successMessage = nil
            }
            
            return .success(())
        } catch {
            return .failure(error)
        }
    }
    
    /**
     * Get friend by ID
     */
    func getFriendById(_ id: String) -> Friend? {
        return friends.first { $0.id == id }
    }
    
    /**
     * Check if friend is currently online
     */
    func isFriendOnline(_ friendId: String) -> Bool {
        return getFriendById(friendId)?.isOnline ?? false
    }
    
    /**
     * Clear error message
     */
    func clearError() {
        error = nil
    }
    
    /**
     * Clear success message
     */
    func clearSuccessMessage() {
        successMessage = nil
    }
    
    /**
     * Show invite friends interface
     */
    func showInviteFriends() {
        // TODO: Implement invite friends functionality
        print("Show invite friends interface")
    }
    
    /**
     * Handle network connectivity changes
     */
    func onNetworkConnectivityChanged(isConnected: Bool) {
        self.isConnected = isConnected
        
        if isConnected {
            // Restart sync when network is restored
            startRealTimeSync()
        } else {
            error = "No internet connection. Some features may not work."
        }
    }
    
    /**
     * Handle app lifecycle changes
     */
    func onAppResumed() {
        // Refresh data when app comes to foreground
        refreshFriends()
    }
    
    func onAppPaused() {
        // Optionally pause real-time updates to save battery
        // stopRealTimeSync()
    }
    
    // MARK: - Private Methods
    
    /**
     * Setup observers for friends data
     */
    private func setupObservers() {
        observeFriendsData()
    }
    
    /**
     * Observe friends data from Firebase with real-time updates
     */
    private func observeFriendsData() {
        // Observe all friends
        friendsRepository.getFriends()
            .receive(on: DispatchQueue.main)
            .sink(
                receiveCompletion: { [weak self] completion in
                    if case .failure(let error) = completion {
                        self?.handleError(error)
                    }
                },
                receiveValue: { [weak self] friends in
                    self?.handleFriendsUpdate(friends)
                }
            )
            .store(in: &cancellables)
    }
    
    /**
     * Handle friends data update
     */
    private func handleFriendsUpdate(_ allFriends: [Friend]) {
        friends = allFriends
        onlineFriends = allFriends.filter { $0.isOnline }
        isLoading = false
        isConnected = true
        lastUpdateTime = Date()
        
        print("Friends updated: \(allFriends.count) total, \(onlineFriends.count) online")
    }
    
    /**
     * Handle errors
     */
    private func handleError(_ error: Error) {
        self.error = "Failed to load friends: \(error.localizedDescription)"
        isLoading = false
        isConnected = false
        
        print("Error loading friends: \(error)")
    }
}

// MARK: - Extensions

extension Friend {
    var isOnline: Bool {
        // Implement online status logic based on your Friend model
        // This is a placeholder implementation
        return friendshipStatus == .accepted && locationSharingEnabled
    }
    
    var statusText: String {
        if isOnline {
            return "Online now"
        } else {
            // You could implement more sophisticated status text based on last seen time
            return "Offline"
        }
    }
}